File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        r_dict = {}
        m_dict = {}
        
        for r in ransomNote:
            r_dict[r] = r_dict.get(r, 0) + 1
        
        for m in magazine:
            m_dict[m] = m_dict.get(m, 0) + 1
        
        for key, value in r_dict.items():
            if key not in m_dict or m_dict[key] < value:
                return False
        
        return True

File: 0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < magazine.length(); i++) {
            map.put(magazine.charAt(i), map.getOrDefault(magazine.charAt(i), 0) + 1);
        }

        for (int i = 0; i < ransomNote.length(); i++) {
            char c = ransomNote.charAt(i);
            int cnt = map.getOrDefault(c, 0);
            if (cnt == 0) {
                return false;
            }

            map.put(c, --cnt);
        }

        return true;
    }
}